### PR TITLE
make unicorn use the physical addresses

### PR DESCRIPTION
**Pull request by PhilippTakacs**
This allows to use unicorn for full system emulation. The emulated code can completely use the MMU and change the page tables.

For better use there is also an hack to emulate non full system programs on x86. This more or less map virtual addresses directly to the physical addresses when the MMU is disabled. I'm not sure if such hacks are also required for other platforms.

### CODE REVIEW

**Comment by wtdcode** [comment]
1. The target branch is `dev` branch, please make the pr mode readable.
2. I can't understand the changes of 60ccb0d248e1b10202eb8e297a0087f45b20a0ac . Note you are changing some core features of Unicorn, please add some tests/samples to illustrate better.
3. The commit 24e9e3c5954a37ebca22c133df931124376e2d61 is too hacky. At least you should support all architectures unicorn currently offers. But still I don't think it's a __correct__ way to do so. I have a copy of integrating qemu usermode with Unicorn but need more time to publish and you might wait for it.

**Comment by PhilippTakacs** [comment]
> I can't understand the changes of https://github.com/unicorn-engine/unicorn/commit/60ccb0d248e1b10202eb8e297a0087f45b20a0ac

The basic changes are add the physical address to the tlb entry and move the memory hooks after the tlb lookup.

I'll add some tests. I might copy the PoC in #1718.

> The commit https://github.com/unicorn-engine/unicorn/commit/24e9e3c5954a37ebca22c133df931124376e2d61 is too hacky.

This changes only under which condition the currently implemented hack is used. Before this change the TLB lookups are done but ignored so couldn't use the MMU.

> At least you should support all architectures unicorn currently offers.

I don't think such hacks are required for other architectures, because there MMUs are not so complex as the x86/amd64 MMU. I currently looking at the code and the definitions of the other architectures.

> But still I don't think it's a correct way to do so. I have a fork of integrating qemu usermode with Unicorn but need more time to polish and you might wait for it.

A simpler way would be to implement a virtual TLB. So you could choose between the TLB implementation of the architecture and this virtual TLB. The virtual TLB defaults to a 1:1 mapping but allow to change entries.

**Comment by PhilippTakacs** [comment]
First test for arm64 is there more will come.

for the hacks: I haven't seen any additional points for hacks yet. But I'm not finished with the checks.

**Comment by PhilippTakacs** [comment]
The amd64 test illustrate why I would like this change (and later a virtual TLB). To emulate multiple processes and shared memory it's necessary to have more access to the TLB.

I'm currently searching, if there are other places where register writes have side effects.

**Comment by PhilippTakacs** [comment]
Hi

I have found that other also might need some changes that you can control the special registers over unicorn, but I think this is a different issue.

I believe my patch works on all architectures and current work on tests for the other architectures.

It would be nice to get some more feedback on this PR.

**Comment by wtdcode** [comment]
> Hi
> 
> I have found that other also might need some changes that you can control the special registers over unicorn, but I think this is a different issue.
> 
> I believe my patch works on all architectures and current work on tests for the other architectures.
> 
> It would be nice to get some more feedback on this PR.

Thanks for your work and sorry I'm still busy these days. I think i could review the whole thing next week or so.

File: qemu/target/i386/excp_helper.c [hunk 1]
```diff
@@ -635,19 +635,19 @@ static int handle_mmu_fault(CPUState *cs, vaddr addr, int size,
     /* align to page_size */
     pte &= PG_ADDRESS_MASK & ~(page_size - 1);
     page_offset = addr & (page_size - 1);
-    paddr = get_hphys(cs, pte + page_offset, is_write1, &prot);
+    /* HACK allow full 64 bit mapping in u64 without paging */
+    if (env->cr[0] & CR0_PG_MASK) {
+        paddr = get_hphys(cs, pte + page_offset, is_write1, &prot);
+    } else {
+        paddr = addr;
```
Comment (wtdcode) [inline qemu/target/i386/excp_helper.c:642]: It looks like not congruent with the previous logic? At least we should have `paddr = vaddr = addr & TARGET_PAGE_MASK`.

File: qemu/accel/tcg/cputlb.c [hunk 2]
```diff
@@ -1436,7 +1438,28 @@ load_helper(CPUArchState *env, target_ulong addr, TCGMemOpIdx oi,
     bool handled;
     HOOK_FOREACH_VAR_DECLARE;
     struct uc_struct *uc = env->uc;
-    MemoryRegion *mr = memory_mapping(uc, addr);
+
+    /* Handle CPU specific unaligned behaviour */
+    if (addr & ((1 << a_bits) - 1)) {
```
Comment (wtdcode) [inline qemu/accel/tcg/cputlb.c:1443]: Any reason to bring this forward?

**Review by wtdcode — COMMENTED** [review]
I start to understand your motivation and I like the current work! See my comments above.

File: qemu/target/i386/excp_helper.c (hunk 1 above)
Comment (PhilippTakacs) [reply qemu/target/i386/excp_helper.c:642]: This is already done direct before the tlb_set_page_with_attrs() call (line 648). But if you like it better I can add this.

File: qemu/include/exec/cpu-defs.h [hunk 3]
```diff
@@ -77,9 +77,9 @@ typedef uint64_t target_ulong;
 #define CPU_VTLB_SIZE 8
 
 #if HOST_LONG_BITS == 32 && TARGET_LONG_BITS == 32
-#define CPU_TLB_ENTRY_BITS 4
-#else
 #define CPU_TLB_ENTRY_BITS 5
+#else
+#define CPU_TLB_ENTRY_BITS 6
```
[the remark this answers is not in the document]
Comment (PhilippTakacs) [reply qemu/include/exec/cpu-defs.h:82]: I have add the `paddr` to the `struct CPUTLBEntry`, which has a `padding to get a power of two size`. To get this padding right this has to be increased.

File: tests/regress/arm64_mmu.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
+#include "unicorn/unicorn.h"
```
[the remark this answers is not in the document]
Comment (PhilippTakacs) [reply tests/regress/arm64_mmu.c]: I have moved the tests, a sample will come

File: qemu/accel/tcg/cputlb.c (hunk 2 above)
Comment (PhilippTakacs) [reply qemu/accel/tcg/cputlb.c:1444]: I have split up the definition and the initialization, if you meant something else please explain.

**Comment by PhilippTakacs** [comment]
I have noticed, that `uc_ctl_flush_tlb` does only clear the tb mappings and not the tlb. Should I create an extra pr for this or add a fix to this pr?

File: tests/regress/arm64_mmu.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (PhilippTakacs) [reply tests/regress/arm64_mmu.c]: I have add some simple sample code. Do you want more tests or sample code?

**Comment by PhilippTakacs** [comment]
Hi

Just wanted to ask if you still miss some changes/fixes or you haven't had time to look again at this pr?

**Comment by wtdcode** [comment]
> Hi
> 
> Just wanted to ask if you still miss some changes/fixes or you haven't had time to look again at this pr?

Sorry for late. I think I could find some time next week.

**Comment by PhilippTakacs** [comment]
Just wanted to ask when you have time to look again at this pr?

**Comment by wtdcode** [comment]
> Just wanted to ask when you have time to look again at this pr?

Sorry I'm just too busy recently but I will try to get it done these days.

**Comment by wtdcode** [comment]
@PhilippTakacs Any reason for the force push?

**Comment by PhilippTakacs** [comment]
> @PhilippTakacs Any reason for the force push?

I have cleaned up some conflicts. These conflicts are not related to this PR. I believe it is better to rebase the PR to current dev. This keeps the changes clean from other changes and makes the git history better to understand.

**Comment by PhilippTakacs** [comment]
close for #1746

**Pull request closed without merging**